### PR TITLE
Rewrite cmd/linkcheck

### LINKED ISSUE #15916: Documentation 404s
Lots of links in the documentation currently lead to 404s. There are even more in the current [published](http://kubernetes.io/v1.0/docs/) version.

In addition to fixing these, we should either set up a test or a pre-commit hook to check that URLs are valid.

Hacky regex, not including relative URLs:
```console
$ for MATCH in $(ag -o 'https?://[^) "\]\n]+[^) "\]\n.]' docs -G ".*\.md$"); do URL=$(echo $MATCH | sed 's/.*\(http.*\)/\1/'); if [ $(curl -s -o /dev/null -w "%{http_code}" $URL) -eq "404" ]; then echo $MATCH; fi; done
docs/user-guide/kubectl/kubectl_replace.md:47:https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/definitions.html
docs/user-guide/replication-controller.md:127:https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/definitions.html#_v1_replicationcontroller
docs/user-guide/pods.md:167:https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/definitions.html#_v1_pod
docs/user-guide/services.md:590:https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/definitions.html#_v1_service
docs/user-guide/pod-states.md:58:https://godoc.org/k8s.io/kubernetes/pkg/api/v1#Probe
docs/user-guide/pod-states.md:58:https://godoc.org/k8s.io/kubernetes/pkg/api/v1#Handler
docs/user-guide/pod-states.md:77:https://godoc.org/k8s.io/kubernetes/pkg/api/v1#PodStatus
docs/user-guide/pod-states.md:77:https://godoc.org/k8s.io/kubernetes/pkg/api/v1#ContainerState
docs/user-guide/deploying-applications.md:79:https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/definitions.html#_v1_replicationcontroller
docs/user-guide/connecting-applications.md:131:https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/definitions.html#_v1_service
docs/user-guide/configuring-containers.md:112:https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/definitions.html#_v1_pod
docs/user-guide/images.md:93:https://www.googleapis.com/auth/devstorage.read_only`,
docs/user-guide/kubectl/kubectl_patch.md:45:https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/definitions.html
docs/getting-started-guides/coreos/bare_metal_offline.md:243:https://storage.googleapis.com/kubernetes-release/release/v0.15.0/bin/linux/amd64/kubecfg
docs/getting-started-guides/coreos/bare_metal_offline.md:246:https://storage.googleapis.com/k8s/flanneld
docs/devel/pull-requests.md:55:https://github.com/kubernetes/contrib/tree/master/submit-queue
docs/devel/pull-requests.md:62:https://github.com/kubernetes/contrib/tree/master/submit-queue/whitelist.txt
docs/devel/automation.md:49:https://github.com/kubernetes/contrib/tree/master/submit-queue
docs/admin/node.md:260:https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/definitions.html#_v1_node
```

### PR DESCRIPTION
**Pull request by caesarxuchao**
Fix #15916, #16233.

Now the linkcheck actually visits a url and checks the response code. It reports url as invalid if the status code >=400.

This check requires a long time to run. Checking docs/ folder alone takes around 5 minutes. Some possible improvements:
1. only run the check on the diff of a PR;
2. add a whitelist of valid URLs, and update the whitelist periodically.

Any suggestions?
cc @lavalamp @nikhiljindal @bgrant0607 @timstclair

### CODE REVIEW

**Comment by caesarxuchao** [comment]
I'll make it a cron job that runs on Jenkins once a day.

**Comment by caesarxuchao** [comment]
First bad link found by the tool:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/v1beta1/types.go#L313

**Comment by caesarxuchao** [comment]
@nikhiljindal A friendly ping.

**Comment by nikhiljindal** [comment]
`hack/verify-godeps.sh` is failing with `Your godeps changes are not reproducable`.
@caesarxuchao Can you please take a look?

File: cmd/linkcheck/links.go [hunk 1]
```diff
@@ -14,39 +14,47 @@ See the License for the specific language governing permissions and
...
 */
 
-//This command checks if the hyperlinks in files are valid. It checks the files
-//with 'fileSuffix' in 'rootDir' for URLs that match 'prefix'. It trims the
-//'prefix' from the URL, uses what's left as the relative path to repoRoot to
-//verify if the link is valid. For example:
-//$ linkcheck --root-dir=${TYPEROOT} --repo-root=${KUBE_ROOT} \
-//  --file-suffix=types.go --prefix=http://releases.k8s.io/HEAD
+// This tool extracts the links from types.go and .md files, visits the link and
+// check the status code of the response.
```
Comment (nikhiljindal) [inline cmd/linkcheck/links.go]: check -> checks

File: cmd/linkcheck/links.go [hunk 2]
```diff
@@ -14,39 +14,47 @@ See the License for the specific language governing permissions and
...
 
 var (
-	httpRE *regexp.Regexp
-
 	rootDir    = flag.String("root-dir", "", "Root directory containing documents to be processed.")
-	repoRoot   = flag.String("repo-root", "", `Root directory of k8s repository.`)
-	fileSuffix = flag.String("file-suffix", "", "suffix of files to be checked")
-	prefix     = flag.String("prefix", "", "Longest common prefix of the link URL, e.g., http://release.k8s.io/HEAD/ for links in pkg/api/types.go")
+	fileSuffix = flag.StringSlice("file-suffix", []string{"types.go", ".md"}, "suffix of files to be checked")
+	whiteList  = []*regexp.Regexp{
```
Comment (nikhiljindal) [inline cmd/linkcheck/links.go]: Add a comment saying what is means to be in this whitelist?
When should someone add a link to this list?

File: cmd/linkcheck/links.go [hunk 3]
```diff
@@ -14,39 +14,47 @@ See the License for the specific language governing permissions and
...
+		regexp.MustCompile(`https://kubernetes-site\.appspot\.com`),
+		regexp.MustCompile(`http://www\.apache\.org/licenses/LICENSE-2\.0`),
+		regexp.MustCompile(`http://kubernetes\.io/img/warning\.png`),
+		regexp.MustCompile(`http://releases\.k8s\.io$`),
+		// skip url that doesn't start with an English alphabet.
+		regexp.MustCompile(`https?://[^A-Za-z].*`),
```
Comment (nikhiljindal) [inline cmd/linkcheck/links.go]: Why this?
Do we have such URLs?

File: cmd/linkcheck/links.go [hunk 4]
```diff
@@ -14,39 +14,47 @@ See the License for the specific language governing permissions and
...
+	}
+	htmlpreviewReg = regexp.MustCompile(`https://htmlpreview\.github\.io/\?`)
+	httpOrhttpsReg = regexp.MustCompile(`https?.*`)
 )
 
-func newWalkFunc(invalidLink *bool) filepath.WalkFunc {
+func newWalkFunc(invalidLink *bool, fileSuffixTable map[string]struct{}, client *http.Client) filepath.WalkFunc {
 	return func(filePath string, info os.FileInfo, err error) error {
-		if !strings.HasSuffix(info.Name(), *fileSuffix) {
+		_, found1 := fileSuffixTable[filepath.Ext(filePath)]
```
Comment (nikhiljindal) [inline cmd/linkcheck/links.go]: Its not clear what are you trying to find here.
Maybe use better variable name than found1 and found2?

File: cmd/linkcheck/links.go [hunk 5]
```diff
@@ -54,16 +62,52 @@ func newWalkFunc(invalidLink *bool) filepath.WalkFunc {
 			return err
 		}
 		foundInvalid := false
-		matches := httpRE.FindAllSubmatch(fileBytes, -1)
-		for _, match := range matches {
-			//match[1] should look like docs/devel/api-conventions.md
-			if _, err := os.Stat(path.Join(*repoRoot, string(match[1]))); err != nil {
-				fmt.Fprintf(os.Stderr, "Link is not valid: %s\n", string(match[0]))
+		allURL := xurls.Strict.FindAll(fileBytes, -1)
```
Comment (nikhiljindal) [inline cmd/linkcheck/links.go]: allURL -> allURLs

File: cmd/linkcheck/links.go [hunk 6]
```diff
@@ -54,16 +62,52 @@ func newWalkFunc(invalidLink *bool) filepath.WalkFunc {
...
+			}
+			for _, whiteURL := range whiteList {
+				if whiteURL.Match(URL) {
+					continue URL
+				}
+			}
+			// remove the htmlpreview Prefix
+			processedURL := htmlpreviewReg.ReplaceAll(URL, []byte{})
+			retry := 0
+			const maxRetry int = 3
```
Comment (nikhiljindal) [inline cmd/linkcheck/links.go]: optional: Make this a flag?

File: cmd/linkcheck/links.go [hunk 7]
```diff
@@ -54,16 +62,52 @@ func newWalkFunc(invalidLink *bool) filepath.WalkFunc {
...
+			for retry < maxRetry {
+				fmt.Fprintf(os.Stdout, "Visiting %s\n", string(processedURL))
+				resp, err := client.Head(string(processedURL))
+				// We only check response code >= 400. We don't care about
+				// errors.
+				if err != nil {
+					break
+				}
+				if resp.StatusCode == 429 {
+					fmt.Fprintf(os.Stderr, "Got %d visiting %s, retry after %d seconds.\n", resp.StatusCode, string(URL), backoff)
```
Comment (nikhiljindal) [inline cmd/linkcheck/links.go]: The right way is to read the "Retry-after" header which should tell you the number of seconds you should wait before making another call.

File: cmd/linkcheck/links.go [hunk 8]
```diff
@@ -54,16 +62,52 @@ func newWalkFunc(invalidLink *bool) filepath.WalkFunc {
...
+			retry := 0
+			const maxRetry int = 3
+			backoff := 100
+			for retry < maxRetry {
+				fmt.Fprintf(os.Stdout, "Visiting %s\n", string(processedURL))
+				resp, err := client.Head(string(processedURL))
+				// We only check response code >= 400. We don't care about
+				// errors.
+				if err != nil {
+					break
```
Comment (nikhiljindal) [inline cmd/linkcheck/links.go]: Should you maybe retry? Silently ignoring doesnt seem right

File: hack/after-build/verify-linkcheck.sh [hunk 9]
```diff
@@ -24,15 +24,20 @@ source "${KUBE_ROOT}/hack/lib/init.sh"
...
-  exit 1
-fi
+APIROOT="${KUBE_ROOT}/pkg/api/"
+APISROOT="${KUBE_ROOT}/pkg/apis/"
+DOCROOT="${KUBE_ROOT}/docs/"
+ROOTS=($APIROOT $APISROOT $DOCROOT)
+for root in "${ROOTS[@]}"; do
+  "${linkcheck}" "--root-dir=${root}" && ret=0 || ret=$?
+  if [[ $ret -eq 1 ]]; then
+    echo "links in ${root} is out of date."
```
Comment (nikhiljindal) [inline hack/after-build/verify-linkcheck.sh]: Shouldnt it be "found invalid links in ${root}"

File: hooks/pre-commit [hunk 10]
```diff
@@ -112,28 +112,6 @@ else
...
-  echo "${red}ERROR!"
-  echo "Some links in pkg/api/.*types.go are outdated. They require a manual fix."
-  exit_code=1
-else
-  echo "${green}OK"
-fi
-echo "${reset}"
-
-echo -ne "Checking for docs that need updating... "
-if ! hack/after-build/verify-generated-docs.sh > /dev/null; then
```
Comment (nikhiljindal) [inline hooks/pre-commit]: Why removing this?

**Comment by nikhiljindal** [comment]
Thanks! Added a few comments.

cc @mvdan who might be interested in our usage of github.com/mvdan/xurls

File: cmd/linkcheck/links.go (hunk 3 above)
Comment (caesarxuchao) [inline cmd/linkcheck/links.go]: In examples, there are url to a special host's IP address, we need to skip those.

File: cmd/linkcheck/links.go (hunk 4 above)
Comment (caesarxuchao) [inline cmd/linkcheck/links.go]: We want to check is file is *.md or types.go, one is an extension, the other one is a file name, so we need to search twice. I'll add a comment.

File: cmd/linkcheck/links.go (hunk 8 above)
Comment (caesarxuchao) [inline cmd/linkcheck/links.go]: Some error I noted down, they are cuased by fake URLs:
```
	// Unexpected error: Get http://server:port: dial tcp: unknown port tcp/port
	// Unexpected error: Get http://fed-master:4001/v2/keys/coreos.com/network/subnets: dial tcp: lookup fed-master: no such host
	// Unexpected error: Get http://kube-master:8080/api/v1: dial tcp: lookup kube-master: no such host
	// Unexpected error: Get http://gitserver.$(POD_NAMESPACE):$(SERVICE_PORT): dial tcp: unknown port tcp/$(SERVICE_PORT)
	// Unexpected error: Get http://ip:port: dial tcp: unknown port tcp/port
```

I'll check again tomorrow, if the number of such links is small, I'll just add them to the whiteList.

File: cmd/linkcheck/links.go (hunk 7 above)
Comment (caesarxuchao) [inline cmd/linkcheck/links.go]: Oh, thanks! Let me try it out.

File: hack/after-build/verify-linkcheck.sh (hunk 9 above)
Comment (caesarxuchao) [inline hack/after-build/verify-linkcheck.sh]: Good catch. Thanks.

File: hooks/pre-commit (hunk 10 above)
Comment (caesarxuchao) [inline hooks/pre-commit]: A slip. I will add it back. Thanks.

**Comment by caesarxuchao** [comment]
I addressed most comments and left an note on why we'd better keep ignoring the error.

One more thing we need to fix before merging: our munger generates `The latest release of this document can be found here.` no matter if the doc actually exists in the latest release, the link checker reports many such errors.

**Comment by nikhiljindal** [comment]
@caesarxuchao Whats the status of this?

**Comment by caesarxuchao** [comment]
It's in my todo list this week :)

**Comment by caesarxuchao** [comment]
The output of hack/verify-linkcheck.sh, all found invalid links are really invalid:
```
Summary of invalid links:
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/types.go, Got 404 visiting http://releases.k8s.io/HEAD/docs/admin/daemon.md
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/docs/design/event_compression.md, Got 404 visiting http://releases.k8s.io/HEAD/pkg/client/unversioned/record/event.go
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/docs/getting-started-guides/coreos/bare_metal_offline.md, Got 404 visiting https://storage.googleapis.com/kubernetes-release/release/v0.15.0/bin/linux/amd64/kubecfg
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/docs/getting-started-guides/coreos/bare_metal_offline.md, Got 404 visiting https://storage.googleapis.com/k8s/flanneld
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/docs/getting-started-guides/logging.md, Got 404 visiting http://releases.k8s.io/release-1.1/contrib/logging/fluentd-sidecar-gcp/README.md
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/docs/getting-started-guides/rackspace.md, Got 404 visiting https://github.com/kubernetes/kubernetes/archive/v0.9.tar.gz
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/docs/getting-started-guides/ubuntu-calico.md, Got 404 visiting https://github.com/kubernetes/kubernetes/releases/download/v1.1.0/kubernetes.tar.gz
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/docs/user-guide/production-pods.md, Got 404 visiting http://releases.k8s.io/HEAD/contrib/git-sync/
```

**Comment by caesarxuchao** [comment]
@nikhiljindal, PTAL. Thanks!

**Comment by mvdan** [comment]
Thanks for the ping @nikhiljindal!

@caesarxuchao what do you mean by this?

    // xurls is confused by http://storage.googleapis.com/kubernetes-release/release/v${K8S_VERSION}/bin/darwin/amd64/kubectl.

**Comment by caesarxuchao** [comment]
Hi @mvdan, I mean when parsing "http://storage.googleapis.com/kubernetes-release/release/v${K8S_VERSION}/bin/darwin/amd64/kubectl.", xurls identifies ""http://storage.googleapis.com/kubernetes-release/release/v$" as a url, rather than the whole string, so I add the partial string into the white list.

xurls is very easy to use, thanks for providing it!

File: cmd/linkcheck/links.go [hunk 11]
```diff
@@ -14,56 +14,158 @@ See the License for the specific language governing permissions and
...
+	// URLs listed in the fullURLWhiteList won't be checked. This separated from
+	// the RegWhiteList to improve efficiency. This list includes dummy URLs that
+	// are hard to be generalized by a regex, and URLs that will cause false alerts.
+	fullURLWhiteList = map[string]struct{}{
+		"http://github.com/some/repo.git": {},
+		// This URL returns 404 when visited by this tool, but it works fine if visited by a browser.
+		"http://stackoverflow.com/questions/ask?tags=kubernetes":  {},
+		"https://github.com/$YOUR_GITHUB_USERNAME/kubernetes.git": {},
+		"https://github.com/$YOUR_GITHUB_USERNAME/kubernetes":     {},
+		// xurls identifies the part before "{" in URL "http://storage.googleapis.com/kubernetes-release/release/v${K8S_VERSION}/bin/darwin/amd64/kubectl" as an URL.
```
Comment (caesarxuchao) [inline cmd/linkcheck/links.go]: @mvdan, I reworded the comment to be more specific.

**Comment by mvdan** [comment]
@caesarxuchao I have added support for braces in https://github.com/mvdan/xurls/commit/1b768d7c393abd8e8dda1458385a57becd4b2d4e, this should fix it for you.

**Comment by caesarxuchao** [comment]
@mvdan, yes, that's fixed. Thank you!

**Comment by nikhiljindal** [comment]
Added a few minor comments to https://github.com/caesarxuchao/kubernetes/commit/12c1a1910cece8690efb79781c7d7e7a5d054959.

**Comment by caesarxuchao** [comment]
@nikhiljindal I pushed another commit to address the comments. PTAL. Thanks.

File: cmd/linkcheck/links.go [hunk 12]
```diff
@@ -14,56 +14,155 @@ See the License for the specific language governing permissions and
...
-	fileSuffix = flag.String("file-suffix", "", "suffix of files to be checked")
-	prefix     = flag.String("prefix", "", "Longest common prefix of the link URL, e.g., http://release.k8s.io/HEAD/ for links in pkg/api/types.go")
+	fileSuffix = flag.StringSlice("file-suffix", []string{"types.go", ".md"}, "suffix of files to be checked")
+	// URLs matching the patterns in the regWhiteList won't be checked. Patterns
+	// of dummy URLs should be added to the list to avoid false alerts. Also,
+	// patterns of URLs that we don't care about can be added here to improve
+	// efficiency.
+	regWhiteList = []*regexp.Regexp{
+		regexp.MustCompile(`https://kubernetes-site\.appspot\.com`),
+		// skip url that doesn't start with an English alphabet.
```
Comment (nikhiljindal) [inline cmd/linkcheck/links.go]: Say in the comment that this is for URLs with IP addresses.

File: Makefile [hunk 13]
```diff
@@ -49,7 +49,6 @@ verify:
 	hack/verify-generated-deep-copies.sh
 	hack/verify-generated-docs.sh
 	hack/verify-swagger-spec.sh
-	hack/verify-linkcheck.sh
```
Comment (nikhiljindal) [inline Makefile]: Why this change?

File: cmd/linkcheck/links.go (hunk 12 above)
Comment (caesarxuchao) [inline cmd/linkcheck/links.go]: Done.

File: Makefile (hunk 13 above)
Comment (caesarxuchao) [inline Makefile]: This link tool will not be run in Jenkins unit-test or e2e builds, it will have its own Jenkins project which runs once every day, because checking all URLs takes too much time.

**Comment by caesarxuchao** [comment]
@nikhiljindal, I addressed your comments. Could you take another look? Thanks.

**Comment by nikhiljindal** [comment]
LGTM, thanks!

**Comment by nikhiljindal** [comment]
Feel free to self apply the LGTM label after you have squashed the commits.

**Comment by caesarxuchao** [comment]
Apply the LGTM per nikhil's comment.

**Comment by ixdy** [comment]
@spxtr can you help @caesarxuchao get `kubernetes-check-links` set up again with a proper checked-in Jenkins config now that this is ready to be checked in?

**Comment by spxtr** [comment]
Sure.

@caesarxuchao If you want this to run regularly on Jenkins you'll need to add a job config for it in `hack/jenkins/job-configs`. You can probably copy most of `kubernetes-test-go`'s config, but you won't need to use a template. See [Jenkins Job Builder](http://docs.openstack.org/infra/jenkins-job-builder/) for docs, and bother me with questions.

**Comment by caesarxuchao** [comment]
A straight-forward rebase, adding back the label.

**Comment by alex-mohr** [comment]
@k8s-bot test this

**Comment by caesarxuchao** [comment]
travis failure:
```
Your godeps changes are not reproducable
diff -Naupr '--exclude=symlink' /home/travis/gopath/src/github.com/kubernetes/kubernetes/Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt /tmp/gopath.4eSANu/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt
--- /home/travis/gopath/src/github.com/kubernetes/kubernetes/Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt	1970-01-01 00:00:00.000000000 +0000
+++ /tmp/gopath.4eSANu/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt	2016-01-21
```

I vaguely remember there are changes to this licence verification. I'll try if rebase solve the problem.

**Comment by caesarxuchao** [comment]
@karlkfi do you have any idea why `hack/verify-godep-license.sh` complains about Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt? The repo doesn't have that file and this PR doesn't touch that folder.

**Comment by caesarxuchao** [comment]
@k8s-bot e2e test this please

**Comment by caesarxuchao** [comment]
#19977 should have fixed the godep error.

**Comment by caesarxuchao** [comment]
Just rebase, so adding back the lgtm.
